### PR TITLE
CLN: remove ABCIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
It's a footgun.

BTW I expect the CI to fail until #38020 goes in, which will [demonstrate](https://github.com/pandas-dev/pandas/pull/38020#discussion_r529159757) that there is a test case that uses MultiIndex.

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -53,10 +53,20 @@ def _join_i8_wrapper(joinf, with_indexers: bool = True):
...
     def wrapper(left, right):
-        if isinstance(left, (np.ndarray, ABCIndex, ABCSeries, DatetimeLikeArrayMixin)):
+        # Note: these only get called with left.dtype == right.dtype
+        if isinstance(
+            left, (np.ndarray, DatetimeIndexOpsMixin, ABCSeries, DatetimeLikeArrayMixin)
+        ):
             left = left.view("i8")
-        if isinstance(right, (np.ndarray, ABCIndex, ABCSeries, DatetimeLikeArrayMixin)):
+        else:
+            raise TypeError(type(left))
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimelike.py:62]: woops, didnt mean for this to get committed

**Comment by jreback** [comment]
need a rebase.

should consider renaming ABCIndexClass -> ABCIndex

**Comment by jreback** [comment]
lgtm ping on green.

**Comment by jbrockmendel** [comment]
ping

**Comment by simonjayhawkins** [comment]
nice. also helps for consistency when casting for mypy

**Comment by simonjayhawkins** [comment]
> should consider renaming ABCIndexClass -> ABCIndex

+1

**Pull request merged**